src/main/java/com/oath/halodb/HaloDBInternal.java

```diff
@@ -281,6 +288,99 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
         }
     }
 
+    //TODO: use fine-grained lock if possible
+    public synchronized boolean takeSnapshot() {
+        logger.info("Start generating the snapshot");
+
+        if (this.isTombstoneFilesMerging) {
+            logger.error("DB is not ready for snapshot now");
+            return false;
+        }
+
+        try {
+            final int currentWriteFileId;
+            compactionManager.pauseCompactionThread();
+
+            // Only support one snapshot now
+            // TODO: support multiple snapshots if needed
+            File snapshotDir = getSnapshotDirectory();
+            if (snapshotDir.exists()) {
+                logger.warn("The snapshot dir is already existed. Delete the old one.");
+                FileUtils.deleteDirectory(snapshotDir);
+            }
+
+            FileUtils.createDirectoryIfNotExists(snapshotDir);
+            logger.info("Created directory for snapshot {}", snapshotDir.toString());
+
+            writeLock.lock();
+            // TODO: Need solutions to handle the file pieces after the rollover
+            // especially there are frequently rollovers
+            try {
+                forceRollOverCurrentWriteFile();
+                currentTombstoneFile = forceRollOverTombstoneFile(currentTombstoneFile);
+
+                currentWriteFileId = currentWriteFile.getFileId();
+            } catch (IOException e) {
+                logger.warn("IO exception when rollover current write files", e);
```

Comment:
Do we need to catch exception here? Let it thrown out?

Reply:
This exception could be handled inside the HaloDB. It would return the false which means the snapshot failed.